### PR TITLE
EXOGTN-2384 : reset PicketLink transaction status on commit and rollback

### PR DESCRIPTION
**Pull request by thomasdelhomenie**
When using the option **lazyStartOfHibernateTransaction**, PicketLink keeps its own Hibernate transaction status (besides the one already kept by Hibernate).
This status is hold in a ThreadLocal and can be one of these values:
* **null** - when no transaction has been started or asked to be started
* **false** - when a transaction has been asked to be started but is not started yet (lazy start)
* **true** - when the transaction is started

This status is well reset to null when a transaction is committed or rollbacked successfully, but not when one of these operations fail.
In such a case, the status is still set to true. Therefore when a new request uses the same thread (threads are generally reused from a pool), the ThreadLocal is already set to true whereas the underlying Hibernate transactin is not started.
When a transaction start is asked in PicketLink, it checks the transaction status, and since it sees the status as 'started' (true), it does not start a new transaction. Then when committing or rollbacking, it does on a non-started transaction, thus the exception "Transaction not successfully started".

This transaction status management is done in a PicketLink class. Therefore this class must be copied into eXo codebase to allow to change it. It has been copied in org.picketlink.idm.impl.store.hibernate.ExoHibernateIdentityStoreSessionImpl and used in the class org.picketlink.idm.impl.store.hibernate.PatchedHibernateIdentityStoreImpl.
The fix always resets the PicketLink Hibernate transaction status to null in the ThreadLocal after a commit or a rollback, to be sure to have a consistent status.
Two tests have been added:
* one to check that the transaction is well managed when an error is raised during a commit. This is done thanks to Byteman, which allows to inject code (throwing an exception in our case) in classes.
* one to check that the rollback is correctly performed when an error is raised before committing the transaction. This test is only done for Hibernate only configuration (no LDAP), that's why it is done in a dedicated test class.

### CODE REVIEW

**Comment by fdrouet** [comment]
> Therefore this class must be copied into eXo codebase to allow to change it.

To help the reviewers, the copied class is https://github.com/picketlink/picketlink-idm/blob/1.4.5.Final/picketlink-idm-hibernate/src/main/java/org/picketlink/idm/impl/store/hibernate/HibernateIdentityStoreSessionImpl.java

File: component/identity/src/main/java/org/picketlink/idm/impl/store/hibernate/ExoHibernateIdentityStoreSessionImpl.java [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+  public void commitTransaction() {
+    if (lazyStartOfHibernateTransaction) {
+      Boolean hbTxStatus = hibernateTxStatus.get();
+
+      // Commit hibernate transaction only if it has really been started
+      if (hbTxStatus != null && hbTxStatus) {
+        try {
+          commitHibernateTransaction();
+        } finally {
+          hibernateTxStatus.set(null);
```
Comment (fdrouet) [inline component/identity/src/main/java/org/picketlink/idm/impl/store/hibernate/ExoHibernateIdentityStoreSessionImpl.java]: For later understanding perhaps it worth a comment / reference to the jira issue

File: component/identity/src/main/java/org/picketlink/idm/impl/store/hibernate/ExoHibernateIdentityStoreSessionImpl.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+  public void rollbackTransaction() {
+    if (lazyStartOfHibernateTransaction) {
+      Boolean hbTxStatus = hibernateTxStatus.get();
+      // Rollback hibernate transaction only if it has really been started
+      if (hbTxStatus != null && hbTxStatus) {
+        try {
+          rollbackHibernateTransaction();
+        } finally {
+          hibernateTxStatus.set(null);
```
Comment (fdrouet) [inline component/identity/src/main/java/org/picketlink/idm/impl/store/hibernate/ExoHibernateIdentityStoreSessionImpl.java]: For later understanding perhaps it worth a comment / reference to the jira issue

File: component/identity/src/main/java/org/picketlink/idm/impl/store/hibernate/ExoHibernateIdentityStoreSessionImpl.java [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+      // Commit hibernate transaction only if it has really been started
+      if (hbTxStatus != null && hbTxStatus) {
+        try {
+          commitHibernateTransaction();
+        } finally {
+          hibernateTxStatus.set(null);
+        }
+      }
+    } else {
+      commitHibernateTransaction();
```
Comment (fdrouet) [inline component/identity/src/main/java/org/picketlink/idm/impl/store/hibernate/ExoHibernateIdentityStoreSessionImpl.java]: Why don't we need to set the hibernate transaction to null in that case ?

File: component/identity/src/main/java/org/picketlink/idm/impl/store/hibernate/ExoHibernateIdentityStoreSessionImpl.java [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+      // Rollback hibernate transaction only if it has really been started
+      if (hbTxStatus != null && hbTxStatus) {
+        try {
+          rollbackHibernateTransaction();
+        } finally {
+          hibernateTxStatus.set(null);
+        }
+      }
+    } else {
+      rollbackHibernateTransaction();
```
Comment (fdrouet) [inline component/identity/src/main/java/org/picketlink/idm/impl/store/hibernate/ExoHibernateIdentityStoreSessionImpl.java]: Why don't we need to set the hibernate transaction to null in that case ?

File: component/identity/src/main/java/org/picketlink/idm/impl/store/hibernate/ExoHibernateIdentityStoreSessionImpl.java (hunk 3 above)
Comment (trongtt) [reply component/identity/src/main/java/org/picketlink/idm/impl/store/hibernate/ExoHibernateIdentityStoreSessionImpl.java]: I see the `hibernateTxStatus` is only useful for `lazyStartOfHibernateTransaction == true`

File: component/identity/src/main/java/org/picketlink/idm/impl/store/hibernate/ExoHibernateIdentityStoreSessionImpl.java (hunk 4 above)
Comment (trongtt) [reply component/identity/src/main/java/org/picketlink/idm/impl/store/hibernate/ExoHibernateIdentityStoreSessionImpl.java]: Same here, the hibernateTxStatus is only useful for lazyStartOfHibernateTransaction == true

**Comment by trongtt** [comment]
I have got a following error while building the project, not sure if this is just my local issue:

> [ERROR] Failed to execute goal on project exo.portal.component.identity: Could not resolve dependencies for project org.exoplatform.gatein.portal:exo.portal.component.identity:jar:5.2.x-SNAPSHOT: Could not find artifact com.sun:tools:jar:1.6 at specified path /working/tools/jdk1.8.0_60.jdk/Contents/Home/jre/../Classes/classes.jar -> [Help 1]

File: component/identity/src/main/java/org/picketlink/idm/impl/store/hibernate/ExoHibernateIdentityStoreSessionImpl.java (hunk 1 above)
Comment (thomasdelhomenie) [reply component/identity/src/main/java/org/picketlink/idm/impl/store/hibernate/ExoHibernateIdentityStoreSessionImpl.java]: ok

File: component/identity/src/main/java/org/picketlink/idm/impl/store/hibernate/ExoHibernateIdentityStoreSessionImpl.java (hunk 2 above)
Comment (thomasdelhomenie) [reply component/identity/src/main/java/org/picketlink/idm/impl/store/hibernate/ExoHibernateIdentityStoreSessionImpl.java]: ok

File: component/identity/src/main/java/org/picketlink/idm/impl/store/hibernate/ExoHibernateIdentityStoreSessionImpl.java (hunk 3 above)
Comment (thomasdelhomenie) [reply component/identity/src/main/java/org/picketlink/idm/impl/store/hibernate/ExoHibernateIdentityStoreSessionImpl.java]: Right. When lazyStartOfHibernateTransaction is false, we rely directly on the Hibernate transaction status, we do not manage a status on PicketLink side (in hibernateTxStatus).

**Comment by thomasdelhomenie** [comment]
@trongtt I cannot reproduce this issue, but I see that some maven config related to classes.jar have been added in social project when using byteman. I will check. It is maybe related to mac only.

File: component/identity/src/main/java/org/picketlink/idm/impl/store/hibernate/ExoHibernateIdentityStoreSessionImpl.java (hunk 1 above)
Comment (thomasdelhomenie) [reply component/identity/src/main/java/org/picketlink/idm/impl/store/hibernate/ExoHibernateIdentityStoreSessionImpl.java]: done

File: component/identity/src/main/java/org/picketlink/idm/impl/store/hibernate/ExoHibernateIdentityStoreSessionImpl.java (hunk 2 above)
Comment (thomasdelhomenie) [reply component/identity/src/main/java/org/picketlink/idm/impl/store/hibernate/ExoHibernateIdentityStoreSessionImpl.java]: done

**Pull request merged**